sdc/native/hashmap.hpp

```diff
@@ -0,0 +1,907 @@
+}
+
+template <typename key_type>
+void delete_generic_value_hashmap(void* p_hash_map)
+{
+
+    auto p_hash_map_spec = (generic_value_hashmap<key_type>*)p_hash_map;
+    for (auto kv_pair: (*p_hash_map_spec->p_map)) {
+        p_hash_map_spec->val_info.decref(kv_pair.second);
+        free(kv_pair.second);
+    }
+    delete p_hash_map_spec;
+}
+
+void delete_generic_hashmap(void* p_hash_map)
+{
+    auto p_hash_map_spec = (generic_hashmap*)p_hash_map;
+    for (auto kv_pair: (*p_hash_map_spec->p_map)) {
+        p_hash_map_spec->key_info.decref(kv_pair.first);
+        free(kv_pair.first);
+        p_hash_map_spec->val_info.decref(kv_pair.second);
+        free(kv_pair.second);
+    }
+    delete p_hash_map_spec;
+}
+
+template <typename key_type, typename value_type>
+void delete_numeric_hashmap(void* p_hash_map)
+{
+
+    auto p_hash_map_spec = (numeric_hashmap<key_type, value_type>*)p_hash_map;
+    delete p_hash_map_spec;
+}
+
+
+template <typename key_type, typename value_type>
+void delete_iter_state(void* p_iter_state)
+{
+    auto p_iter_state_spec = reinterpret_cast<iter_state*>(p_iter_state);
+    auto p_hash_map_spec = reinterpet_hashmap_ptr<key_type, value_type>(p_iter_state_spec->second);
+    using itertype = typename std::remove_reference<decltype(*p_hash_map_spec)>::type::iterator_type;
+    auto p_hash_map_iter = reinterpret_cast<itertype*>(p_iter_state_spec->first);
+
+    delete p_hash_map_iter;
+    delete p_iter_state_spec;
+}
+
+
+template<typename Key,
+         typename HashCompare
+>
+void GenericHashmapType<Key, void*, HashCompare>::set(Key key, void* val)
+{
+    auto vsize = this->val_info.size;
+    void* _val = malloc(vsize);
+    memcpy(_val, val, vsize);
+
+    typename map_type::value_type inserted_node(key, _val);
+    {
+        typename map_type::accessor existing_node;
+        bool ok = this->p_map->insert(existing_node, inserted_node);
+        if (ok)
+        {
+            // isnertion succeeded need to incref value
+            this->val_info.incref(val);
+        }
+        else
+        {
+            // insertion failed key already exists
+            this->val_info.decref(existing_node->second);
+            free(existing_node->second);
+            existing_node->second = _val;
+            this->val_info.incref(val);
+        }
+    }
+}
+
+template<typename Key,
+         typename HashCompare
+>
+void GenericHashmapType<Key, void*, HashCompare>::clear()
+{
+    for (auto kv_pair: (*this->p_map)) {
+        this->val_info.decref(kv_pair.second);
+        free(kv_pair.second);
+    }
+    this->p_map->clear();
+}
+
+
+template<typename Key,
+         typename HashCompare
+>
+int8_t GenericHashmapType<Key, void*, HashCompare>::pop(Key key, void* res) {
+    bool found = false;
+    {
+        typename map_type::const_accessor result;
+        found = this->p_map->find(result, key);
+        if (found)
+        {
+            memcpy(res, result->second, this->val_info.size);
```

Comment:
Clokwork will aggro on `memcpy`

Reply:
I double checked and it appears it doesn't (see report below - all three issues are known and should be fixed by #967). Not sure if `memcpy_s` will enhance something, but I can still change if you think it's better.

```
Klocwork 20.1.0.97
-----
May 4, 2021 2:01:17 PM CDT: Running build preparation stage...
May 4, 2021 2:01:17 PM CDT: Build preparation stage completed
May 4, 2021 2:01:18 PM CDT: Running compilation stage...
1> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/_distributed.cpp
2> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/_hiframes.cpp
9> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/native/sort.cpp
8> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/transport/hpat_transport_single_process.cpp
7> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/io/_io.cpp
6> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/_datetime_ext.cpp
4> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/_str_ext.cpp
3> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/_set_ext.cpp
Warning: /localdisk/work/akozlov/miniconda3/kwrun_6/.kwlp/buildspec.txt:24: 'compile' line for object file /localdisk/work/akozlov/miniconda3/hpat/build/temp.linux-x86_64-3.8/sdc/native/utils.o is already defined (ignored)
15> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/native/utils.cpp
14> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/native/module.cpp
11> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/native/stable_sort.cpp
5> CLASSIC: Compiling /localdisk/work/akozlov/miniconda3/hpat/sdc/native/conc_dict_module.cpp
May 4, 2021 2:01:29 PM CDT: Compilation stage completed
May 4, 2021 2:01:29 PM CDT: Running C/C++ defects detection stage...
1> Analyzing _distributed.cpp_0.o
2> Analyzing _hiframes.cpp_1.o
3> Analyzing _set_ext.cpp_2.o
4> Analyzing _str_ext.cpp_3.o
5> Analyzing _datetime_ext.cpp_4.o
6> Analyzing _io.cpp_5.o
7> Analyzing hpat_transport_single_process.cpp_6.o
8> Analyzing utils.cpp_10.o
9> Analyzing sort.cpp_7.o
10> Analyzing stable_sort.cpp_8.o
11> Analyzing module.cpp_9.o
12> Analyzing conc_dict_module.cpp_12.o
May 4, 2021 2:02:21 PM CDT: C/C++ defects detection stage completed
May 4, 2021 2:02:21 PM CDT: Running linking stage...
May 4, 2021 2:02:22 PM CDT: Linking stage completed
1 (Local) /localdisk/work/akozlov/miniconda3/hpat/sdc/native/sort.cpp:151 VOIDRET (2:Error) Analyze
void function returns value

2 (Local) /localdisk/work/akozlov/miniconda3/hpat/sdc/native/utils.cpp:89 NPD.CHECK.MUST (1:Critical) Analyze
Pointer 'this->arena._M_t' checked for NULL at line 86 will be dereferenced at line 89.

3 (Local) /localdisk/work/akozlov/miniconda3/hpat/sdc/native/stable_sort.cpp:340 VOIDRET (2:Error) Analyze
void function returns value

Summary: 3 Local
3 Total Issue(s)
```